File: utils/test-utils/src/main/java/de/sovity/edc/extension/e2e/junit/utils/InstancesForEachConnector.java
/*
 * Copyright (c) 2024 sovity GmbH
 *
 *  This program and the accompanying materials are made available under the
 *  terms of the Apache License, Version 2.0 which is available at
 *  https://www.apache.org/licenses/LICENSE-2.0
 *
 *  SPDX-License-Identifier: Apache-2.0
 *
 *  Contributors:
 *       sovity GmbH - initial API and implementation
 *
 */

package de.sovity.edc.extension.e2e.junit.utils;

import org.junit.jupiter.api.extension.ExtensionContext;
import org.junit.jupiter.api.extension.ParameterContext;
import org.junit.jupiter.api.extension.ParameterResolutionException;
import org.junit.jupiter.api.extension.ParameterResolver;

import java.util.List;
import java.util.Map;
import java.util.function.BiFunction;

import static java.util.function.Function.identity;
import static java.util.stream.Collectors.toMap;


public class InstancesForEachConnector<S> implements ParameterResolver {
    private final List<S> sides;
    private final BiFunction<ParameterContext, ExtensionContext, S> getSideOrNull;
    private final Map<S, InstancesForJunitTest> instances;

    public InstancesForEachConnector(List<S> sides, BiFunction<ParameterContext, ExtensionContext, S> getSideOrNull) {
        this.sides = sides;
        this.getSideOrNull = getSideOrNull;
        instances = sides.stream()
            .collect(toMap(identity(), unused -> new InstancesForJunitTest()));
    }

    public <T> List<T> all(Class<T> clazz) {
        return sides.stream().flatMap(side -> forSide(side).all(clazz).stream()).toList();
    }

    public InstancesForJunitTest forSide(S side) {
        var instancesOfSide = instances.get(side);
        if (instancesOfSide == null) {
            throw new IllegalArgumentException("No instances for side %s found".formatted(side));
        }
        return instancesOfSide;
    }

    @Override
    public boolean supportsParameter(
        ParameterContext parameterContext,
        ExtensionContext extensionContext
    ) throws ParameterResolutionException {
        var side = getSideOrNull.apply(parameterContext, extensionContext);
        if (side == null) {
            return false;
        }

        return forSide(side).supportsParameter(parameterContext, extensionContext);
    }

    @Override
    public Object resolveParameter(
        ParameterContext parameterContext,
        ExtensionContext extensionContext
    ) throws ParameterResolutionException {
        var side = getSideOrNull.apply(parameterContext, extensionContext);
        if (side == null) {
            throw new ParameterResolutionException("No side found for parameter %s".formatted(parameterContext.getParameter()));
        }

        return forSide(side).resolveParameter(parameterContext, extensionContext);
    }
}

File: connector/ce/libs/runtime-test-lib/src/main/java/de/sovity/edc/extension/e2e/junit/utils/InstancesForEachConnector.kt
/*
 * Copyright sovity GmbH and/or licensed to sovity GmbH under one or
 * more contributor license agreements. You may not use this file except
 * in compliance with the "Elastic License 2.0".
 *
 * SPDX-License-Identifier: Elastic-2.0
 */
package de.sovity.edc.extension.e2e.junit.utils

import org.junit.jupiter.api.extension.ExtensionContext
import org.junit.jupiter.api.extension.ParameterContext
import org.junit.jupiter.api.extension.ParameterResolutionException
import org.junit.jupiter.api.extension.ParameterResolver

/**
 * A fancy [Map] of `S` and [InstancesForJunitTest].
 *
 * @param S Enum for "Provider, Consumer, ProviderDp, ConsumerDp, etc."
 */
class InstancesForEachConnector<S>(
    private val sides: List<S>,
    private val getSideOrNull: (parameter: ParameterContext, extension: ExtensionContext?) -> S?
) : ParameterResolver {
    private val instances = sides.associateWith { InstancesForJunitTest() }

    fun <T> all(clazz: Class<T>): List<T> = sides
        .map { forSide(it) }
        .filter { it.has(clazz) }
        .map { it.get(clazz) }

    fun forSide(side: S): InstancesForJunitTest = instances[side]
        ?: error("No instances for side $side found")

    override fun supportsParameter(
        parameterContext: ParameterContext,
        extensionContext: ExtensionContext?
    ): Boolean {
        val side = getSideOrNull(parameterContext, extensionContext)
        if (side == null) {
            return false
        }
        return forSide(side).supportsParameter(parameterContext, extensionContext)
    }

    override fun resolveParameter(
        parameterContext: ParameterContext,
        extensionContext: ExtensionContext?
    ): Any? {
        val side = getSideOrNull(parameterContext, extensionContext)
        if (side == null) {
            throw ParameterResolutionException("No side found for parameter ${parameterContext.parameter}")
        }
        return forSide(side).resolveParameter(parameterContext, extensionContext)
    }
}
